File: LeetCode75/Find_All_Anagrams_In_a_String/Find_All_Anagrams_In_a_String.cc
// 438. Fina All Anagrams in a String C++
#include <vector>
#include <string>
#include <unordered_map>

class Solution {
public:
    std::vector<int> findAnagrams(std::string s, std::string p) {
        std::unordered_map<char, int> mpP;
        std::unordered_map<char, int> mpS;
        std::vector<int> output;
        for (auto c: p) mpP[c]++;
        int j = p.size();
        for (int i = 0; i < s.size(); i++) {
            mpS[s[i]]++;
            if (i >= j) {
                mpS[s[i-j]] == 1 ? mpS.erase(s[i-j]) : mpS[s[i-j]]--;
            }
            if (mpP == mpS) {
                output.emplace_back(i-j+1);
            }
        }
        return output;
    }
};

File: LeetCode75_LevelOne/Find_All_Anagrams_In_a_String/Find_All_Anagrams_In_a_String.py
# 438. Fina All Anagrams in a String Python
from collections import Counter

class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        dic = Counter(p)
        
        output = []
        i = 0
        j = len(p)
        dicTwo = Counter(s[:j])
        while j <= len(s):
            if dic == dicTwo:
                output.append(i)
            dicTwo[s[i]] -= 1
            if dicTwo[s[i]] <= 0:
                dicTwo.pop(s[i])
            if j < len(s):
                dicTwo[s[j]] += 1
            i += 1
            j += 1
        return output
